# Class 1

In [1]:
import statistics

ages = [19, 22, 21, 35, 28, 24, 31, 19, 27, 40]

# a. Number of people
number_of_people = len(ages)
print("Number of people:", number_of_people)

# b. Total of all ages
total_age = sum(ages)
print("Total of all ages:", total_age)

# c. Mean age
mean_age = total_age / number_of_people
print("Mean age:", round(mean_age, 1))

# d. Youngest and oldest ages
youngest = min(ages)
oldest = max(ages)
print("Youngest age:", youngest)
print("Oldest age:", oldest)

# e. Median
median_age = statistics.median(ages)
print("Median age:", median_age)

# Compare median and mean
print("Is the median different from the mean?", median_age != mean_age)

Number of people: 10
Total of all ages: 266
Mean age: 26.6
Youngest age: 19
Oldest age: 40
Median age: 25.5
Is the median different from the mean? True


# Class 2

In [1]:
import pandas as pd

data = pd.read_csv("50_Startups.csv")

print(data.shape)

(50, 5)


In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   R&D Spend        50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing Spend  50 non-null     float64
 3   State            50 non-null     object 
 4   Profit           50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


none of the columns have missing values, all entries are non-null so no missing value treatment is needed.

In [3]:
# As a Series
series = data["Profit"]

# As a single-column DataFrame
dataframe = data[["Profit"]]

print(type(series))
print(series.shape)

print(type(dataframe))
print(dataframe.shape)

<class 'pandas.core.series.Series'>
(50,)
<class 'pandas.core.frame.DataFrame'>
(50, 1)


key seperating difference is the double brackets

In [4]:
data.describe()

,R&D Spend,Administration,Marketing Spend,Profit
count,50.000000,50.000000,50.000000,50.000000
mean,73721.615600,121344.639600,211025.097800,112012.639200
std,45902.256482,28017.802755,122290.310726,40306.180338
min,0.000000,51283.140000,0.000000,14681.400000
25%,39936.370000,103730.875000,129300.132500,90138.902500
50%,73051.080000,122699.795000,212716.240000,107978.190000
75%,101602.800000,144842.180000,299469.085000,139765.977500
max,165349.200000,182645.560000,471784.100000,192261.830000


mean score = 112012.639
regarding "out of place", there are some unusually high and low values in the "spend" column

In [6]:
data.to_json(orient="records")

data.to_json("50_Startups.json", orient="records")

Each row became a single JSON record

# Class 3

In [11]:
import pandas as pd

data = pd.read_excel("50_startups_messy.xlsx")

print(data.shape)
print(data.isnull().sum())
print(data.duplicated().sum())

(50, 5)
R&D Spend          0
Administration     0
Marketing Spend    1
State              0
Profit             0
dtype: int64
0


In [12]:
# Drop duplicate rows
data = data.drop_duplicates()

# Standardise the State column
data["State"] = data["State"].str.strip().str.title()

# Convert Marketing Spend to numeric
data["Marketing Spend"] = pd.to_numeric(data["Marketing Spend"], errors="coerce")

print(data)

    R&D Spend  Administration  Marketing Spend       State     Profit
0   165349.20       136897.80        471784.10    New York  192261.83
1   162597.70       151377.59        443898.53  California  191792.06
2   153441.51       101145.55        407934.54     Florida  191050.39
3   144372.41       118671.85        383199.62    New York  182901.99
4   142107.34        91391.77        366168.42     Florida  166187.94
5   131876.90        99814.71        362861.36    New York  156991.12
6   134615.46       147198.87        127716.82  California  156122.51
7   130298.13       145530.06        323876.68     Florida  155752.60
8   120542.52       148718.95              NaN    New York  152211.77
9   123334.88       108679.17        304981.62  California  149759.96
10  101913.08       110594.11        229160.95     Florida  146121.95
11  100671.96        91790.61        249744.55  California  144259.40
12   93863.75       127320.38        249839.44     Florida  141585.52
13   91992.39       

In [13]:
simple_sample = data.sample(n=20, random_state=7)

print(simple_sample)

    R&D Spend  Administration  Marketing Spend       State     Profit
13   91992.39       135495.07        252664.93  California  134307.35
15  114523.61       122616.84        261776.23    New York  129917.04
22   73994.56       122782.75        303319.26     Florida  110352.25
42   23640.93        96189.63        148001.11  California   71498.49
30   61994.48       115641.28         91131.24     Florida   99937.59
27   72107.60       127864.55              NaN    New York  105008.31
1   162597.70       151377.59        443898.53  California  191792.06
20   76253.86       113867.30        298664.47  California  118474.03
10  101913.08       110594.11        229160.95     Florida  146121.95
36   28663.76       127056.21        201126.82     Florida   90708.19
32   63408.86       129219.61         46085.25  California   97427.84
46    1315.46       115816.21        297114.46     Florida   49490.75
34   46426.07       157693.92        210797.67  California   96712.80
18   91749.16       

In [14]:
stratified_sample = data.groupby("State", group_keys=False).sample(
    frac=0.20,
    random_state=7
)

print(stratified_sample)

    R&D Spend  Administration  Marketing Spend       State     Profit
42   23640.93        96189.63        148001.11  California   71498.49
39   38558.51        82982.09        174999.30  California   81005.76
49       0.00       116983.80         45173.06  California   14681.40
7   130298.13       145530.06        323876.68     Florida  155752.60
23   67532.53       105751.03        304768.73     Florida  108733.99
4   142107.34        91391.77        366168.42     Florida  166187.94
17   94657.16       145077.58        282574.31    New York  125370.37
45    1000.23       124153.04          1903.93    New York   64926.08
31   61136.38       152701.92         88218.23    New York   97483.56


In [16]:
print("Population means:")
print(data[["R&D Spend", "Marketing Spend"]].mean())

print("\nSimple random sample means:")
print(simple_sample[["R&D Spend", "Marketing Spend"]].mean())

print("\nStratified sample means:")
print(stratified_sample[["R&D Spend", "Marketing Spend"]].mean())

Population means:
R&D Spend           73721.615600
Marketing Spend    207414.790851
dtype: float64

Simple random sample means:
R&D Spend           76076.374500
Marketing Spend    239540.931053
dtype: float64

Stratified sample means:
R&D Spend           62103.467778
Marketing Spend    192853.752222
dtype: float64


the stratified sample is closer because its means have smaller differences from the population means